lib/ai/model/tools/auditquery.go

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2023 Gravitational, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package tools
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sashabaranov/go-openai"
+
+	"github.com/gravitational/teleport/gen/go/eventschema"
+	"github.com/gravitational/teleport/lib/ai/tokens"
+)
+
+type AuditQueryGenerationTool struct {
+	LLM *openai.Client
+}
+
+func (t *AuditQueryGenerationTool) Name() string {
+	return "Audit Query Generation"
+}
+
+func (t *AuditQueryGenerationTool) Description() string {
+	return `Generates a SQL query that can be ran against teleport audit events.
+The input must be a single string describing what the query must achieve.`
+}
+
+func (t *AuditQueryGenerationTool) Run(_ context.Context, _ *ToolContext, _ string) (string, error) {
+	// This is stubbed because AuditQueryGenerationTool is handled specially.
+	// This is because execution of this tool breaks the loop and returns a command suggestion to the user.
+	// It is still handled as a tool because testing has shown that the LLM behaves better when it is treated as a tool.
+	//
+	// In addition, treating it as a Tool interface item simplifies the display and prompt assembly logic significantly.
+	return "", trace.NotImplemented("not implemented")
+}
+
+func (t *AuditQueryGenerationTool) ChooseEventTable(ctx context.Context, input string, tc *tokens.TokenCount) (string, error) {
+	tableList, err := eventschema.QueryableEventList()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	prompt := []openai.ChatCompletionMessage{
+		{
+			Role: openai.ChatMessageRoleSystem,
+			Content: `You are a tool that find the correct table to run a query on.
```

Comment:
This sounds a bit odd 😅 
```suggestion
			Content: `Your job it to find the correct table to run a query on.
```